### ISSUE TITLE
Remove numeric_cast from `image_util.cpp`

### ISSUE THREAD

**Issue by springmeyer** [mapnik/mapnik#2893]
`image_util.cpp` using `numeric_cast` in performance critical places `fill` mostly but also `get_pixel`. This needs to be removed to avoid code bloat from exception handling in functions that should otherwise be simple and easy for the compiler to inline.

**Comment by artemp** [comment]
@springmeyer @flippmoke 

`fill` is less problematic because `numeric_cast<>` is called only once before applying result to all pixels.

'get/set_pixel' are potentially performance issue even so boost docs suggest that implementation should not incur runtime overhead when input doesn't overflow. 

`boost::numeric_cast' is policy based but I don't see any simple ways to implement something like 

``` 
if (range_check == cNegOverflow)  return std::numeric_limits<T>::min() ;
else if (range_check == cPosOverflow) return std::numeric_limits<T>::max() ;
else return static_cast<T>(input);
```

Maybe we can borrow some range checking facilities from `numeric_cast` and implement our own `pixel_cast`. Alternatively we can just use `static_cast<T>`  ??

**Comment by artemp** [comment]
@springmeyer @flippmoke 

I'm going to settle on 
``` c++
template <typename T, typename S>
inline T pixel_cast(S s)
{
    using namespace boost::numeric;
    if (s > bounds<T>::highest() )
    {
        return bounds<T>::highest();
    }
    else if (s < bounds<T>::lowest())
    {
        return bounds<T>::lowest();
    }
    else return static_cast<T>(s);
}
```

**Comment by artemp** [comment]
96f7120ecc11c81ebe90c91d2db9757a844ef2eb - replaces all numeric_cast instances with pixel_cast

**Comment by springmeyer** [comment]
@artemp above plan is great except that `s > bounds<T>::highest()` fails unpredictably when the sign of `S` and `T` differ.

**Comment by springmeyer** [comment]
d29a0f1 should fix the problem - @artemp can you review and merge if it looks good? After that ping @BergWerkGIS to rebuild mapnik packages so we can get unblocked tomorrow on https://github.com/mapnik/node-mapnik/issues/461.

It looks like this was not failing in core because we did not have much test coverage of `mapnik::get_pixel` on anything except colors and not on potentially out of range `set_pixel` values. But @flippmoke added pretty comprehensive tests for this in node-mapnik, and that is what caught this: https://github.com/mapnik/node-mapnik/blob/master/test/image.test.js#L339-L449 :ok_hand:

**Comment by BergWerkGIS** [comment]
I built `mapnik` from branch [fix_pixel_cast](https://github.com/mapnik/mapnik/tree/fix_pixel_cast).
Build was ok, but tests stalled here (I killed `test.exe` after waiting for about 1hr):

![image](https://cloud.githubusercontent.com/assets/4549888/8128658/7898c008-1101-11e5-8447-715fbe7fc922.png)

Then I built `node-mapnik` with the above mapnik sdk.
Build was ok, but tests stalled here (have waited for about 1/2 hr now):

![image](https://cloud.githubusercontent.com/assets/4549888/8128684/b9bc3eb6-1101-11e5-8ceb-11b629c5cde1.png)

**Comment by artemp** [comment]
@BergWerkGIS - could you try master ?

**Comment by BergWerkGIS** [comment]
@artemp 52 errors

```
       "C:\mb\windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.sln" (default target) (1) ->
       "C:\mb\windows-builds-64\packages\mapnik-master\mapnik-gyp\build\csv.vcxproj.metaproj" (default target) (2) ->
       "C:\mb\windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj" (default target) (16) ->
       (ClCompile target) ->
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned char>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\w
       indows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2127: 'max_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned char>::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\w
       indows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2127: 'min_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(135): error C2134: 'std::numeric_limits<dou
       ble>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-
       builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(135): error C2127: 'max_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(134): error C2134: 'std::numeric_limits<dou
       ble>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-
       builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(134): error C2127: 'min_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(152): error C2127: 'max_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(153): error C2127: 'min_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<sig
       ned char>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<sig
       ned char>::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned short>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\
       windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned short>::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\
       windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<sho
       rt>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<sho
       rt>::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned int>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\wi
       ndows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned int>::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\wi
       ndows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<int
       >::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-bui
       lds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<int
       >::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-bui
       lds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(135): error C2134: 'std::numeric_limits<flo
       at>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(134): error C2134: 'std::numeric_limits<flo
       at>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned __int64>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\m
       b\windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned __int64>::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\m
       b\windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<__i
       nt64>::max': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows
       -builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<__i
       nt64>::min': call does not result in a constant expression (compiling source file ..\..\src\image_copy.cpp) [C:\mb\windows
       -builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned int>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\wi
       ndows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2127: 'max_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned int>::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\wi
       ndows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2127: 'min_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(152): error C2127: 'max_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(153): error C2127: 'min_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned char>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\w
       indows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned char>::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\w
       indows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<sig
       ned char>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<sig
       ned char>::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned short>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\
       windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned short>::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\
       windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<sho
       rt>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<sho
       rt>::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<int
       >::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-bui
       lds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<int
       >::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-bui
       lds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(135): error C2134: 'std::numeric_limits<flo
       at>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(135): error C2127: 'max_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(134): error C2134: 'std::numeric_limits<flo
       at>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-b
       uilds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(134): error C2127: 'min_val': illegal initi
       alization of 'constexpr' entity with a non-constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\win
       dows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<uns
       igned __int64>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\m
       b\windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<uns
       igned __int64>::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\m
       b\windows-builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(143): error C2134: 'std::numeric_limits<__i
       nt64>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows
       -builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(142): error C2134: 'std::numeric_limits<__i
       nt64>::min': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows
       -builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(135): error C2134: 'std::numeric_limits<dou
       ble>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-
       builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]
         C:\mb\windows-builds-64\packages\mapnik-master\include\mapnik/pixel_cast.hpp(134): error C2134: 'std::numeric_limits<dou
       ble>::max': call does not result in a constant expression (compiling source file ..\..\src\image_util.cpp) [C:\mb\windows-
       builds-64\packages\mapnik-master\mapnik-gyp\build\mapnik.vcxproj]

    99 Warning(s)
    52 Error(s)
```

**Comment by talaj** [comment]
I got the same, need to add `#include <limits>` somewhere.

**Comment by talaj** [comment]
@BergWerkGIS https://github.com/mapnik/mapnik/pull/2909

**Comment by artemp** [comment]
@talaj - you are using windows ? :D 

a030f5531dc2eecf3d953244b84999b209995813 - added <limits>

**Comment by talaj** [comment]
@artemp no, it was on Debian.

**Comment by BergWerkGIS** [comment]
@artemp re-build failed with similar messages.
I'll try another clean one from scratch.

**Comment by talaj** [comment]
Aha, it seems to be more like here https://github.com/mapnik/mapnik/commit/d3d6b09516293a47f19f9052b112679fa1af9098. That include `<limits>` helped me anyway :-).

**Comment by artemp** [comment]
@talaj - yep ^

In C++11 std::numeric_limits<T>::max()/min() return constexpr  
@BergWerkGIS - looks like windows build needs patching up or ?

**Comment by BergWerkGIS** [comment]
Ran a clean build: same errors.

We also had problems with `constexpr` and `libosimuim`.
So, pataching for Windows might be necessary.

I kicked off a build with latest Visual Studio, to see if it works with more recent version.
Will report back later, as I have to leave now.

**Comment by BergWerkGIS** [comment]
Visual Studio **2015 RC1**:

* `mapnik` build succeeded
* but tests stuck at same place as mentioned above
* `node-mapnik` build succeeded
* but tests stuck at same place as mentioned above

Still, have to find a patch for **VS2014 CTP4**, as this is what we are currently using for our whole stack.

**Comment by springmeyer** [comment]
@BergWerkGIS - thanks for digging! I think I've found the problem and have a fix I'm testing with @artemp that will work with VS2014 CTP4. So, we are unblocked. Have a great weekend!

**Comment by springmeyer** [comment]
Was able to replicate the bug with just:

```cpp
double x = 58.6;
std::int64_t x0 = mapnik::pixel_cast<std::int64_t>(x); // hangs
```

The cause is infinite template recursion as [this](https://github.com/mapnik/mapnik/blob/29b464e868ae883527b8e7bd01738c64ce108930/include/mapnik/pixel_cast.hpp#L105-L107) keeps calling itself. This indicates to me that type deduction is either broken or just works differently on windows such that `greater(double,double)` still matches `greater<double,int64_t>()`.

**Comment by springmeyer** [comment]
Fixed by 8d3d136. See the comments in the commit message of 8d3d136 for details. Now all tests run cleanly and pass on windows (tested with VS 2014).

Take homes:

   - comparing numbers across integer/floating point and signed/unsigned is no small matter. Lots of ways to get it wrong in C/C++. See http://jwwalker.com/pages/safe-compare.html and http://craighenderson.co.uk/papers/numeric_compare/ for why.
   - We've now got some decent code to handle > and < safely for all the types we care about for pixels (which cover a lot). At some point this could be cleaned up and made available elsewhere.
   - That `elsewhere` we should start caring more about: where are comparison across sign/unsigned happening that might also be buggy (mostly on invalid input). Let's get a handle on this over at #2907
   - `numeric_cast` did all this for us before which was nice, but it still makes sense to ban `numeric_cast` in all cases except those where you want an exception when values are out of range.
